hkmc2/shared/src/main/scala/hkmc2/codegen/Deforestation.scala

```
@@ -0,0 +1,1104 @@
+    def addAll(es: Iterable[Symbol -> ProdVar]) = store.addAll(es)
+    def apply(s: Symbol) = store(s)
+  
+  def getClsFields(s: ClassSymbol) = s.tree.clsParams
+
+  
+  
+  def constrain(p: ProdStrat, c: ConsStrat) = constraints ::= p -> c
+  
+  def processBlock(b: Block)(using
+    inArm: LinkedHashMap[ProdVar, ClsOrModSymbol] = LinkedHashMap.empty[ProdVar, ClsOrModSymbol],
+    matching: LinkedHashMap[ResultId, ClsOrModSymbol] = LinkedHashMap.empty[ResultId, ClsOrModSymbol]
+  ): ProdStrat = b match
+    case m@Match(scrut, arms, dflt, rest) =>
+      val scrutStrat = processResult(scrut)
+      constrain(scrutStrat, Dtor(scrut.uid)(m, matching.lastOption.map(_._1))(using this))
+      val armsRes = if arms.forall{ case (cse, _) => cse.isInstanceOf[Case.Cls] } then
+        arms.map:
+          case (Case.Cls(s, _), body) => 
+            processBlock(body)(
+              using inArm + (scrutStrat.asInstanceOf[ProdVar] -> s),
+              matching + (scrut.uid -> s)
+            )
+      else
+        arms.map:
+          case (_, armBody) => processBlock(armBody)
+      val dfltRes = dflt.map(processBlock)
+      rest match
+        case End(msg) =>
+          val matchRes = freshVar()
+          armsRes.appendedAll(dfltRes).foreach: r =>
+            constrain(r, matchRes._2)
+          matchRes._1
+        case _ => processBlock(rest)
+
+    case Return(res, implct) => processResult(res)
+    case Assign(lhs, rhs, rest) =>
+      constrain(processResult(rhs), symToStrat(lhs).asConsStrat)
+      processBlock(rest)
+    case Begin(sub, rest) =>
+      processBlock(sub)
+      processBlock(rest)
+    case Define(defn, rest) =>
+      defn match
+        case FunDefn(_, sym, params, body) =>
+          val funSymStratVar = symToStrat(sym)
+          val param = params.head match
+            case ParamList(flags, params, restParam) => params
+          val funStrat = constrFun(param, body) // TODO: handle mutiple param list
```

If you intend to do it in a later PR, please at least crash here instead of silently doing the wrong thing!! And then add a TODO in a tracking issue for Lumberhack (similar to, eg, https://github.com/hkust-taco/mlscript/issues/263).

---

Thanks for the suggestion. I have updated here so that now it crashes when `restParam` is not `None`, and will open a tracking issue for this.

---

https://github.com/hkust-taco/mlscript/pull/289#discussion_r2063017996